src/core/bpf-firewall.c

```diff
@@ -50,6 +53,256 @@ enum {
+        };
+
+        struct bpf_insn header_insn[] = {
+                /*
+                 * Check if the packet was already accepted by the address check.
+                 * Jump over the port checking if it was. Offset is fixed later.
+                 */
+                BPF_MOV32_REG(BPF_REG_2, BPF_REG_8),
+                BPF_ALU32_IMM(BPF_AND, BPF_REG_2, ACCESS_ALLOWED),
+                BPF_JMP_IMM(BPF_JNE, BPF_REG_2, 0, -1),
+        };
+
+        int r;
+
+        assert(p);
+        assert(map_fd >= 0);
+
+        if (protocol == ETH_P_IP) {
+                int protocol_offset = offsetof(struct iphdr, protocol);
+
+                struct bpf_insn ipv4_insn[] = {
+                        /*
+                         * If skb->protocol != ETH_P_IP, skip this whole block.
+                         * The jump offset is fixed later.
+                         */
+                        BPF_JMP_IMM(BPF_JNE, BPF_REG_7, htobe16(protocol), -1),
+
+                        /*
+                         * Load IP header beginning from the skb into stack. R7
+                         * is the destination buffer in the stack.
+                         */
+                        BPF_MOV64_REG(BPF_REG_1, BPF_REG_6),
+                        BPF_MOV64_IMM(BPF_REG_2, 0),
+                        BPF_MOV64_REG(BPF_REG_7, BPF_REG_10),
+
+                        /*
+                         * Reserve 12 bytes of space in the stack. Then read 12
+                         * bytes to get the values we are interested in: the
+                         * IP header length and the transport layer protocol.
+                         * We do this in one long read instead of two short
+                         * reads.
+                         */
+                        BPF_ALU64_IMM(BPF_ADD, BPF_REG_7, -12),
+                        BPF_MOV64_REG(BPF_REG_3, BPF_REG_7),
+                        BPF_MOV32_IMM(BPF_REG_4, 12),
+                        BPF_RAW_INSN(BPF_JMP | BPF_CALL, 0, 0, 0, BPF_FUNC_skb_load_bytes),
+
+                        /*
+                         * R2 = IHL, header length is (IHL * 32 / 8) bytes. IHL
+                         * bits are located in the first byte (offset 0).
+                         */
+                        BPF_LDX_MEM(BPF_B, BPF_REG_9, BPF_REG_7, 0),
+                        BPF_ALU32_IMM(BPF_AND, BPF_REG_9, 0xf),
+                        BPF_ALU32_IMM(BPF_MUL, BPF_REG_9, 4),
+                        BPF_MOV32_REG(BPF_REG_2, BPF_REG_9),
+
+                        /*
+                         * R9 = transport layer protocol.
+                         */
+                        BPF_LDX_MEM(BPF_B, BPF_REG_9, BPF_REG_7, protocol_offset),
+                };
+
+                /*
+                 * Fix the jumps to the end of the block.
+                 */
+                int prog_len = ELEMENTSOF(header_insn) + ELEMENTSOF(ipv4_insn) + ELEMENTSOF(port_check_insn);
```

Comment:
please use "size_t" for stuff like this that is about sizes of stuff in memory

Reply:
Ok.